runtime/src/main/java/org/corfudb/runtime/object/CorfuCompileProxy.java

```diff
@@ -184,31 +185,35 @@ public CorfuCompileProxy(CorfuRuntime rt, UUID streamID, Class<T> type, Object[]
             }
         }
 
-        // Perform underlying access
-        for (int x = 0; x < rt.getParameters().getTrimRetry(); x++) {
-            // Linearize this read against a timestamp
-            final long timestamp = rt.getSequencerView()
-                            .query(getStreamID());
-            log.debug("Access[{}] conflictObj={} version={}", this, conflictObject, timestamp);
+        // Linearize this read against a timestamp
+        AtomicLong timestamp = new AtomicLong(rt.getSequencerView().query(getStreamID()));
 
-            try {
-                return underlyingObject.access(o -> o.getVersionUnsafe() >= timestamp
-                                && !o.isOptimisticallyModifiedUnsafe(),
-                        o -> o.syncObjectUnsafe(timestamp),
-                        o -> accessMethod.access(o));
-            } catch (TrimmedException te) {
-                log.info("accessInner: Encountered trimmed address space " +
-                    "while accessing version {} of stream {} on attempt {}",
-                    timestamp, getStreamID(), x);
-                // We encountered a TRIM during sync, reset the object
-                underlyingObject.update(o -> {
-                    o.resetUnsafe();
-                    return null;
-                });
-            }
-        }
+        log.debug("Access[{}] conflictObj={} version={}", this, conflictObject, timestamp);
 
-        throw new TrimmedException();
+        // Perform underlying access
+        return underlyingObject.access(o -> o.getVersionUnsafe() >= timestamp.get()
+                        && !o.isOptimisticallyModifiedUnsafe(),
+                o -> {
+                    for (int x = 0; x < rt.getParameters().getTrimRetry(); x++) {
+                        try {
+                            o.syncObjectUnsafe(timestamp.get());
+                            break;
+                        } catch (TrimmedException te) {
+                            log.info("accessInner: Encountered trimmed address space " +
+                                            "while accessing version {} of stream {} on attempt {}",
+                                    timestamp.get(), getStreamID(), x);
+
+                            o.resetUnsafe();
+
+                            if (x == (rt.getParameters().getTrimRetry() - 1)) {
+                                throw te;
+                            }
+
+                            timestamp.set(rt.getSequencerView().query(getStreamID()));
```

Comment:
why do we need to set the timestamp here?  Is it expected to change after it was obtained in line 189?

Reply:
exactly, the state of the log might have changed and you want to provide the latest state.